Auto-generate GraphQL comments in SDL to support a self-documenting API (#5824)
redwoodjs/redwood: docs/docs/graphql.md

  """Fetch Posts."""
  posts: [Post!]!

  """Fetches the Redwood root schema."""
  redwood: Redwood
}

"""
The RedwoodJS Root Schema

Defines details about RedwoodJS such as the current user and version information.
"""
type Redwood {
  """The current user."""
  currentUser: JSON

  """The version of Prisma."""
  prismaVersion: String

  """The version of Redwood."""
  version: String
}

scalar Time

"""Autogenerated input type of UpdateContact."""
input UpdateContactInput {
  """The email of the contact."""
  email: String

  """The message of the contact."""
  message: String

  """The name of the contact."""
  name: String
}

"""Autogenerated input type of UpdatePost."""
input UpdatePostInput {
  """The content of a post."""
  body: String

  """The title of a post."""
  title: String
}
```

#### Root Schema

Documentation is also generated for thr Redwood Root Schema that defines details about Redwood such as the current user and version information.
```
type Query {
  """
  Fetches the Redwood root schema.
  """
  redwood: Redwood
}

"""
The Redwood Root Schema

Defines details about Redwood such as the current user and version information.
"""
type Redwood {

>   """
>   The current user.
>   """

Reviewer: ```suggestion
  """The current user."""
```

Author: I cannot make these changes due to prettier/linting changing them back from one-line to multiline.

  currentUser: JSON

>   """
>   The version of Prisma.
>   """

Reviewer: ```suggestion
  """The version of Prisma."""
```

Author: I cannot make these changes due to prettier/linting changing them back from one-line to multiline.

  prismaVersion: String

>   """
>   The version of Redwood.
>   """

Reviewer: ```suggestion
  """The version of Redwood."""
```

Author: I cannot make these changes due to prettier/linting changing them back from one-line to multiline.

}
```

## Self-Documenting GraphQL API

RedwoodJS helps you document your GraphQL API by auto-generating commented SDL that will be incorporated into the `graphql.schema` file when generating types and that will be used for GraphiQL and the GraphQL Playground explorer.

> If you comment your Prisma schema models, its fields, or enums, the SDL generator will use the comment as the documentation.

Reviewer: ```suggestion
If you comment your Prisma schema models, its fields, or enums, the SDL generator will use those comments as the documentation.
```

If there is no Prisma comment, then the SDL generator will default a comment that you can then edit.

:::note
If you re-generate the SDL, any custom comments will be overwritten.

> However, if you make those edits in your Prisma schema, the those will be used.

Reviewer: ```suggestion
However, if you make those edits in your Prisma schema, then those will be used.
```

* the Model name
* the Enum name
* each Model field name

```
/// A blog post.
model Post {
  /// The unique identifier of a post.
  id        Int      @id @default(autoincrement())
  /// The title of a post.
  title     String
  /// The content of a post.
  body      String
  /// When the post was created.
  createdAt DateTime @default(now())
}

model Contact {
  /// The unique identifier of a contact.
  id        Int @id @default(autoincrement())
  /// The name of the contact.
  name      String
  /// The email of the contact.
  email     String
  /// The message of the contact.
  message   String
  /// When the contact was created.
  createdAt DateTime @default(now())
}

/// A list of allowed colors.
enum Color {
  RED
  GREEN
  BLUE
}
```

### SDL Comments

The SDL generator adds comments for:

* Directives
* Queries
* Mutations
* Input Types

Comments in GraphQL schema are enclosed in `"""`.

```
"""
Use to check whether or not a user is authenticated and is associated
with an optional set of roles.
"""
directive @requireAuth(roles: [String]) on FIELD_DEFINITION

"""Use to skip authentication checks and allow public access."""
directive @skipAuth on FIELD_DEFINITION

> """
> Representation of Contact.
> """

Reviewer: ```suggestion
"""Representation of Contact"""
```

Reviewer: And the same for the rest of the comments that need it

Author: FYi I tested and it reformats when generating, but good idea. 👍

type Contact {
  """
  When the contact was created.
  """
  createdAt: DateTime!

  """
  The email of the contact.
  """
  email: String!

  """
  The unique identifier of a contact.
  """
  id: Int!

  """
  The message of the contact.
  """
  message: String!

  """
  The name of the contact.
  """
  name: String!
}

"""Autogenerated input type of InputContact."""
input CreateContactInput {
  """
  The email of the contact.
  """
  email: String!

  """
  The message of the contact.
  """
  message: String!

  """
  The name of the contact.
  """
  name: String!
}

"""Autogenerated input type of InputPost."""
input CreatePostInput {
  """
  The content of a post.
  """
  body: String!

  """
  The title of a post.
  """
  title: String!
}

"""About mutations"""
type Mutation {
  """Creates a new Contact."""
  createContact(input: CreateContactInput!): Contact!

>   """Creates a new Post."""

Author: I had to update the docs with the change to a one-liner.

}
```

## Self-Documenting GraphQL API

> RedwoodJS helps you document your GraphQL API by auto-generating commented SDL that will be incorporated into the `graphql.schema` file when generating types and that will be used for GraphiQL and the GraphQL Playground explorer.

Author: Correct. Just like it helps you do a crud service if you use `sdl --crud`.

> RedwoodJS helps you document your GraphQL API by auto-generating commented SDL that will be incorporated into the `graphql.schema` file when generating types and that will be used for GraphiQL and the GraphQL Playground explorer.

Reviewer: crud is true by default. docs isn't. Should we maybe clarify that you have to explicitly enable it? So people don't think something's wrong of they don't see any comments. Or get worried they'll get a bunch of generated comments all over their files, depending on what POV they have 😀

> RedwoodJS helps you document your GraphQL API by auto-generating commented SDL that will be incorporated into the `graphql.schema` file when generating types and that will be used for GraphiQL and the GraphQL Playground explorer.

Author: Yes, that intro is clunky. How about?

```
RedwoodJS helps you document your GraphQL API by generating commented SDL used for GraphiQL and the GraphQL Playground explorer and that can be turned into API docs using tools like Docusaurus..

If you specify the SDL generators' `--docs` option, comments (which the [GraphQL spec](https://spec.graphql.org/October2021/#sec-Descriptions) calls "descriptions") will be incorporated into the graphql.schema file when generating types.

If you comment your Prisma schema models, its fields, or enums, the SDL generator will use those comments as the documentation.

If there is no Prisma comment, then the SDL generator will default a comment that you can then edit.
```

> RedwoodJS helps you document your GraphQL API by auto-generating commented SDL that will be incorporated into the `graphql.schema` file when generating types and that will be used for GraphiQL and the GraphQL Playground explorer.

Author: I have added a new intro.

If you comment your Prisma schema models, its fields, or enums, the SDL generator will use those comments as the documentation.

If there is no Prisma comment, then the SDL generator will default a comment that you can then edit.

:::note
If you re-generate the SDL, any custom comments will be overwritten.
However, if you make those edits in your Prisma schema, then those will be used.
:::

### Prisma Schema Comments

> Your Prisma schema is documented with a comment `///` that precedes:

Reviewer: ```suggestion
Your Prisma schema is documented with tripple slash comments (`///`) that precedes:
```

> * the Model name
> * the Enum name

Reviewer: ```suggestion
* Model names
* Enum names
```

* each Model field name

```
/// A blog post.
model Post {
  /// The unique identifier of a post.
  id        Int      @id @default(autoincrement())
  /// The title of a post.
  title     String
  /// The content of a post.
  body      String
  /// When the post was created.
  createdAt DateTime @default(now())
}

model Contact {
  /// The unique identifier of a contact.
  id        Int @id @default(autoincrement())
  /// The name of the contact.
  name      String
  /// The email of the contact.
  email     String
  /// The message of the contact.
  message   String
  /// When the contact was created.
  createdAt DateTime @default(now())

> }

Reviewer: Maybe only keep one of these models in the interest of lowering the "TL;DR" risk :)

Author: Oh, I think I intended to show that the  `Comment` model didn't need a top-level model-level comment.

But, that is confusing.

Removing.

/// A list of allowed colors.
enum Color {
  RED
  GREEN
  BLUE
}
```

### SDL Comments

When used with `--docs` option, [SDL generator](./cli-commands.md#generate-sdl) adds comments for:

* Directives
* Queries
* Mutations
* Input Types

:::note
By default, the `--docs` option to the SDL generator is false and comments are not created.
:::

> Comments in GraphQL schema are enclosed in `"""` and will be included in the generate4d `.redwood/schema.graphql` at the root of your project.

Reviewer: I'm not sure I fully understand this comment. But let me try to write what I think it means, and if I'm right, maybe we can see if we should update it somehow

Comments enclosed in `"""` or `"` in your sdl files will be included in the generated GraphQL schema at the root of your project (`.redwood/schema.graphql`).

Does that sound about right?

```
"""
Use to check whether or not a user is authenticated and is associated
with an optional set of roles.
"""
directive @requireAuth(roles: [String]) on FIELD_DEFINITION

"""Use to skip authentication checks and allow public access."""

> directive @skipAuth on FIELD_DEFINITION

Reviewer: Do you think we could try to make this example shorter? Maybe only include the minimum needed to show off the features you want? Like one short `type`, one short `input` etc

Author: Agreed. I removed the Contact model examples to focus just on Post.

Also, reorganized so while it may not match the generated file, the content is grouped better (scalars, queries/mutations, models, etc)

"""
A blog post.
"""
type Post {
  "The content of a post."
  body: String!

  "Description for createdAt."
  createdAt: DateTime!

  "The unique identifier of a post."
  id: Int!

  "The title of a post."
  title: String!
}

"""
About queries
"""
type Query {
  "Fetch a Contact by id."
  contact(id: Int!): Contact

  "Fetch Contacts."
  contacts: [Contact!]!

  "Fetch a Post by id."
  post(id: Int!): Post

  "Fetch Posts."
  posts: [Post!]!

  "Fetches the Redwood root schema."
  redwood: Redwood
}

"""
The RedwoodJS Root Schema

Defines details about RedwoodJS such as the current user and version information.
"""
type Redwood {
  "The current user."
  currentUser: JSON

  "The version of Prisma."
  prismaVersion: String

  "The version of Redwood."
  version: String
}

scalar Time

"""
Autogenerated input type of UpdateContact.
"""
input UpdateContactInput {
  "The email of the contact."
  email: String

  "The message of the contact."
  message: String

  "The name of the contact."
  name: String
}

"""
Autogenerated input type of UpdatePost.
"""
input UpdatePostInput {
  "The content of a post."
  body: String

  "The title of a post."
  title: String
}
```

#### Root Schema

> Documentation is also generated for thr Redwood Root Schema that defines details about Redwood such as the current user and version information.

Reviewer: ```suggestion
Documentation is also generated for the Redwood Root Schema that defines details about Redwood such as the current user and version information.
```

Author: Fixing typo :)

```
type Query {
  "Fetches the Redwood root schema."
  redwood: Redwood
}

"""
The Redwood Root Schema

Defines details about Redwood such as the current user and version information.
"""
type Redwood {
  "The current user."
  currentUser: JSON

  "The version of Prisma."
  prismaVersion: String

  "The version of Redwood."
  version: String
}

scalar BigInt
scalar Date
scalar DateTime
scalar JSON
scalar JSONObject
scalar Time

```

### Preview in GraphiQL

> The [GraphQL Playground aka GraphiQL](https://www.graphql-yoga.com/docs/features/graphiql) is a way for you to interact with your schema and try out queries and mutations. It can show you the schema by inspecting it. You can find the GraphQL Playground at http://localhost:8911/graphql when your dev server is running.

Reviewer: ```suggestion
The [GraphQL Playground aka GraphiQL](https://www.graphql-yoga.com/docs/features/graphiql) is a way for you to interact with your schema and try out queries and mutations. It can show you the schema by inspecting it. You can find the GraphQL Playground at [http://localhost:8911/graphql](http://localhost:8911/graphql) when your dev server is running.
```

The documentation generated is present when exploring the schema.

#### Queries

> ![graphiql-queries](/img/graphql-api-docs/graphiql-queries.png)

Reviewer: The graphiql screenshots are all rather large. Have you considered resizing them a little bit?

Author: I think the image are small in dimensions but get fit to width in the docs. Let me see if Docusurus has a way of defining a container to set image dimensions.

Author: Solution apparently is to use good o' standard html:

```
<img alt="graphiql-mutations" src="/img/graphql-api-docs/graphiql-mutations.png" width="400" />
```

> #### Mutations
> ![graphiql-mutations](/img/graphql-api-docs/graphiql-mutations.png)
> #### Model Types
> ![graphiql-type](/img/graphql-api-docs/graphiql-type.png)

Reviewer: ```suggestion
#### Mutations
![graphiql-mutations](/img/graphql-api-docs/graphiql-mutations.png)

#### Model Types
![graphiql-type](/img/graphql-api-docs/graphiql-type.png)
```

    ]
  },
```

2. Create a `docs` directory at the root of your project

```terminal
mcd docs
`

3. Install the plugin

```terminal
yarn add @edno/docusaurus2-graphql-doc-generator graphql
```

4. Update `docs/docusaurus.config.js` and configure the plugin and navbar

```
// docs/docusaurus.config.js
// ...
  plugins: [
    [
      '@edno/docusaurus2-graphql-doc-generator',
      {
        schema: '../.redwood/schema.graphql',
        rootPath: './docs',
        baseURL: 'graphql-api',
        homepage: './docs/graphql-api/generated.md',
        linkRoot: '../..',
      },
    ],
  ],
// ...
themeConfig:
    /** @type {import('@docusaurus/preset-classic').ThemeConfig} */
    ({
      navbar: {
        title: 'My Site',
        logo: {
          alt: 'My Site Logo',
          src: 'img/logo.svg',
        },
        items: [
          {
            to: '/docs/graphql-api', // adjust the location depending on your baseURL (see configuration)
            label: 'GraphQL API', // change the label with yours
            position: 'left',
          },
//...
```
5. Update `docs/sidebars.js` to include the generated `graphql-api/sidebar-schema.js`

```
// docs/sidebars.js

> /**
>  * Creating a sidebar enables you to:
>  - create an ordered group of docs
>  - render a sidebar for each doc of that group
>  - provide next/previous navigation
> 
>  The sidebars can be generated from the filesystem, or explicitly defined here.
> 
>  Create as many sidebars as you want.
>  */

Reviewer: ```suggestion
/**
 * Creating a sidebar enables you to:
 *  - create an ordered group of docs
 *  - render a sidebar for each doc of that group
 *  - provide next/previous navigation
 *
 * The sidebars can be generated from the filesystem, or explicitly defined here.
 *
 * Create as many sidebars as you want.
 */
```

// @ts-check

/** @type {import('@docusaurus/plugin-content-docs').SidebarsConfig} */
const sidebars = {
  // By default, Docusaurus generates a sidebar from the docs folder structure
  tutorialSidebar: [
    {
      type: 'autogenerated',
      dirName: '.',
    },
  ],
  ...require('./docs/graphql-api/sidebar-schema.js'),
}

module.exports = sidebars
```

6. Generate the docs

`yarn docusaurus graphql-to-doc`

:::tip
You can overwrite the generated docs and bypass the plugin's diffMethod use `--force`.

``yarn docusaurus graphql-to-doc --force`
:::

7. Start Docusaurus

```
yarn start
```

#### Example Screens

> ##### Schema Documentation
> 
> ![graphql-doc-example-main](/img/graphql-api-docs/schema-doc.png)

Reviewer: ```suggestion
##### Schema Documentation
![graphql-doc-example-main](/img/graphql-api-docs/schema-doc.png)
```
